pgx/src/array.rs

```diff
@@ -0,0 +1,235 @@
+use crate::datum::{Array, FromDatum};
+use crate::pg_sys;
+use core::ptr::{slice_from_raw_parts_mut, NonNull};
+use pgx_pg_sys::*;
+
+extern "C" {
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_DATA_PTR(arrayType: *mut ArrayType) -> *mut u8;
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_DIMS(arrayType: *mut ArrayType) -> *mut libc::c_int;
+    /// # Safety
+    /// Must only be used on a "valid" (Postgres-constructed) ArrayType
+    fn pgx_ARR_NELEMS(arrayType: *mut ArrayType) -> i32;
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_NULLBITMAP(arrayType: *mut ArrayType) -> *mut bits8;
+}
+
+/// Handle describing a bare, "untyped" pointer to a Postgres varlena array,
+/// offering safe accessors to the various fields of the ArrayType,
+/// and access to the various slices that are implicitly present in the varlena.
+///
+/// # On sizes and subscripts
+///
+/// Postgres uses C's `int` (`c_int` in Rust) for sizes, and Rust uses `usize`.
+/// Thus various functions of RawArray return `c_int` values, but you must convert to usize.
+/// On 32-bit or 64-bit machines with 32-bit `c_int`s, you can losslessly upgrade to `as usize`,
+/// except with negative indices, which Postgres asserts against creating.
+/// PGX currently only intentionally supports 64-bit machines,
+/// and while support for ILP32 or I64LP128 C data models may become possible,
+/// PGX will **not** support 16-bit machines in any practical case, even though Rust does.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct RawArray {
+    at: NonNull<ArrayType>,
+}
+
+#[deny(unsafe_op_in_unsafe_fn)]
+impl RawArray {
+    // General implementation notes:
+    // RawArray is not Copy or Clone, making it harder to misuse versus *mut ArrayType.
+    // But this also offers safe accessors to the fields, like &ArrayType,
+    // so it requires validity assertions in order to be constructed.
+    // The main reason it uses NonNull and denies Clone, however, is access soundness:
+    // It is not sound to go from &mut Type to *mut T if *mut T is not a field of Type.
+    // This creates an obvious complication for handing out pointers into varlenas.
+    // Thus also why this does not use lifetime-bounded borrows.
```

Comment:
Talking about the intent of this type might be useful to publicly discuss in its documentation.